### PR TITLE
Update TO Go cdns/:name/federations (CRUDer)

### LINKED ISSUE #1341: Federations APIs: CRUD
- [x] GET /api/$version/cdns/:name/federations
- [x] GET /api/$version/cdns/:name/federations/:fedId
- [x] POST /api/$version/cdns/:name/federations
- [x] PUT /api/$version/cdns/:name/federations/:fedId
- [x] DELETE /api/$version/cdns/:name/federations/:fedId

### LINKED ISSUE #1580: Managing federations on a CDN with no delivery services results in internal server error
For example, this would happen on a brand  new cdn.

I think the problem has to do with an empty IN statement

https://github.com/apache/incubator-trafficcontrol/blob/master/traffic_ops/app/lib/API/Federation.pm#L523

### PR DESCRIPTION
**Pull request by moltzaum**
Converted endpoints:
GET /api/1.2/cdns/:name/federations
GET /api/1.2/cdns/:name/federations/:id
POST /api/1.2/cdns/:name/federations
PUT /api/1.2/cdns/:name/federations/:id
DELETE /api/1.2/cdns/:name/federations/:id

`GetCDNFederationsByName` (a part of the client) hasn't been tested yet because `POST /api/1.2/federations/:id/deliveryservices` isn't implemented, however I did test the functionality using curl and my browser.

fixes #1580
fixes #1341

### CODE REVIEW

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

File: traffic_ops/client/v13/cdnfederations.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+	"github.com/apache/trafficcontrol/lib/go-tc/v13"
+)
+
+/* Internally, the CDNName is only used in the GET method. The CDNName
+ * seems to primarily be used to differentiate between `/federations` and
+ * `/cdns/:name/federations`. Although the behavior is odd, it is kept to
+ * keep the same behavior from perl. */
+
+func (to *Session) CreateCDNFederationByName(f v13.CDNFederation, CDNName string) (*v13.CreateCDNFederationResponse, ReqInf, error) {
```
Comment (rawlinp) [inline traffic_ops/client/v13/cdnfederations.go]: Since these are new client methods, can you please use the Nullable structs in the parameters and responses? I know most of the other client implementations don't follow that pattern, but IMO this is something we need to curb, especially with new client implementations.

The problem is that some fields in the API are optional, and without using the Nullable struct, a client must provide a value for the optional fields (in the case of Federations, it would be "description"). With a Nullable struct, a client could just leave the description field null if they didn't want a description. This also makes it possible to test null values in the API integration tests.

To see what I mean as an example, this is what I did for the Origins endpoint: https://github.com/apache/trafficcontrol/pull/2247/files#diff-c0c1ea4955f245fa2b7861436ec93374R33

Comment (moltzaum) [reply traffic_ops/client/v13/cdnfederations.go]: Is there no case where a non-nullable struct will be used?

Comment (rawlinp) [reply traffic_ops/client/v13/cdnfederations.go]: The API implementation uses the nullable struct, which really only leaves the client implementation. If the client also uses the nullable struct, then there shouldn't be any need for the non-nullable struct.

A lot of the legacy TO Go client code was already using the non-nullable structs, which is why they're still prevalent. IMO we need to go through all the existing client implementations to add methods that use the Nullable structs and properly decode the responses everywhere. We would then deprecate the existing methods that use non-nullable structs and eventually remove them. But for now we should just prevent adding onto that pile of non-nullable struct implementations.

File: lib/go-tc/v13/federations.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+type CDNFederation struct {
+	ID          *int    `json:"id" db:"id"`
+	CName       *string `json:"cname" db:"cname"`
+	Ttl         *int    `json:"ttl" db:"ttl"`
+	Description *string `json:"description" db:"description"`
+
+	//omitempty only works with primitive types and pointers
+	*DeliveryServiceIDs `json:"deliveryService,omitempty"`
+
+	//Extra datapoint for ease in Read function (cannot Scan after StructScan)
```
Comment (rawlinp) [inline lib/go-tc/v13/federations.go]: I'd prefer this to be named `CDNName` rather than just `Name`.

Also can you elaborate on "cannot Scan after StructScan"?

File: traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+		return tc.DBError, tc.SystemError
+	}
+}
+
+//fed.ReqInfo.Params["name"] is not used on creation, rather the cdn name
+//is connected when the federations/:id/deliveryservice links a federation
+//Note: cdns and deliveryservies have a 1-1 relationship
+func (fed *TOCDNFederation) Create() (error, tc.ApiErrorType) {
+
+	if fed.ReqInfo.User.PrivLevel < auth.PrivLevelAdmin {
```
[the remark this answers is not in the document]
Comment (moltzaum) [reply traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: Sounds good. I'll double check to be sure.

File: lib/go-tc/v13/federations.go (hunk 2 above)
Comment (moltzaum) [reply lib/go-tc/v13/federations.go]: A struct scan allows you to scan in such a way that you don't need to scan in order, which works by specifying db tags. After using struct scan, I guess the order cannot be guaranteed, so Scan (which works by order) will not work.

Also, I think a struct scan assumes that everything in the scan queue needs to fit into the struct.

When I had the Nullable and non-nullable structs I think this made a little more sense. Now, I think it'd be best to simply do a sequential scan of all the fields.

Comment (moltzaum) [reply lib/go-tc/v13/federations.go]:
```
//Option 1: (get data, and also fetch cdn information)
SELECT federation.id as id, cname, ttl, description, deliveryservice.id as ds_id, xml_id, cdn.name as cdn_name FROM federation
JOIN federation_deliveryservice as fd ON federation.id = fd.federation
JOIN deliveryservice ON deliveryservice.id = fd.deliveryservice
RIGHT JOIN cdn ON cdn.id = cdn_id

//Option 2: (get data, then check if cdn exists)
SELECT federation.id as id, cname, ttl, description, deliveryservice.id as ds_id, xml_id, cdn.name as cdn_name FROM federation
JOIN federation_deliveryservice as fd ON federation.id = fd.federation
JOIN deliveryservice ON deliveryservice.id = fd.deliveryservice

SELECT id from cdn WHERE cdn.name = :name
```

As far as splitting queries is concerned, would there be a big difference in efficiency? I initially did option 1 because additional database queries could be more expensive, but now I think option 2 would be best, since it would also allow me to scan into the struct directly for the first query.

Comment (rawlinp) [reply lib/go-tc/v13/federations.go]: Yeah, I don't think one extra DB call per GET is really a big deal. Premature optimization is the root of all evil :). Prefer readability/simplicity to start.

Comment (rob05c) [reply lib/go-tc/v13/federations.go]: Why does it need a RIGHT JOIN? The `deliveryservice` table `cdn_id` is NOT NULL, can't you:
```
SELECT federation.id as id, cdn.name as cdn_name, ttl, description, deliveryservice.id as ds_id, xml_id, cdn.name as cdn_name FROM federation
JOIN federation_deliveryservice as fd ON federation.id = fd.federation
JOIN deliveryservice ON deliveryservice.id = fd.deliveryservice
JOIN cdn ON cdn.id = deliveryservice.cdn_id
```
?

I would vote to avoid the StructScan and use a standard `sql.Tx.Scan`, before changing the query. IMO the single query is fast and readable, we shouldn't use suboptimal SQL to accommodate the limitations of the sqlx library. A `Scan(&f.ID, &f.CName, &f.TTL, &s.Description, &s.DeliveryServiceID, &s.XMLID, &cdnID)` isn't that hard to read.

I'm also -1 on putting `json:"-"` tags in `tc` structs, just to accommodate sqlx deficiencies. Our structs are our public interface, for anyone using Go or the Go TO client, as well as all our products that use Go and TO data. We should prioritize keeping that interface clean.

Comment (rawlinp) [reply lib/go-tc/v13/federations.go]: Maybe part of the problem is that we're trying to handle two GET endpoints (`cdns/:name/federations` and `cdns/:name/federations/:fedId`) in the same `Read` handler. Perhaps it would make more sense to split this into two handlers to make them more clear, as the latter doesn't even really care about `cdn_id`.

Comment (moltzaum) [reply lib/go-tc/v13/federations.go]: It doesn't need a right join if you separate out the queries. What I was trying to do was return federation data and cdn data together. That way, I could check to see if the requested cdn existed. Specifying a regular join gets the correct result, but if I got 0 rows I can't tell if thats because the cdn doesn't exist.

Comment (moltzaum) [reply lib/go-tc/v13/federations.go]: BTW, in option 2 the second query would only be run if the first query has 0 rows.

Comment (moltzaum) [reply lib/go-tc/v13/federations.go]: @rawlinp The problem here has to do more with fetching the cdn name. The id can be included or excluded elegantly. All I need to do is conditionally assign a different query (one that I haven't shown), and it should work.

Comment (moltzaum) [reply lib/go-tc/v13/federations.go]: I could just not return a "Resource not found" error inside my handler, but then the behavior would change.

Comment (rob05c) [reply lib/go-tc/v13/federations.go]: >It doesn't need a right join if you separate out the queries.

>if I got 0 rows I can't tell if thats because the cdn doesn't exist.

It doesn't need a RIGHT JOIN as a single query. The cdn must exist: there's a foreign key constraint, and a not-null constraint.

+1 on using different functions. The ID handler can use the CRUDer abstractions, and the name handler can use a native http.HandlerFunc. I believe we have several endpoints already doing that.


Using field mapping like sqlx, and frameworks like ORMs, are always less powerful and expressive than SQL. SQL is expressive, powerful, and readable, as a 4th generation programming language. Likewise, objects aren't relational, and shouldn't be. We shouldn't try to shove a square peg into a round hole, when the data we need doesn't cleanly map to objects. We should use fast, clean, readable SQL; and map that to our objects afterwards. 

The same goes for HTTP/JSON. Clean APIs, clean objects, and clean SQL, with a bit of mapping code where necessary, is better than trying to hack any of those to fit the other. JSON, Objects, and SQL don't map perfectly; but they're all great in their respective contexts. Using each properly in its own context is much better—easier to learn, use, read, write, and generally work with—than trying to hack them to fit each-other, just to avoid a bit of mapping code between them. The mapping code isn't super-pretty, but it's far better than making the API/Object/SQL difficult to use and read everywhere it's used.

But that's just my opinion.

Comment (rob05c) [reply lib/go-tc/v13/federations.go]: Ah, sorry, I didn't understand that your issue was differentiating whether the CDN exists, to return different errors for nonexistent CDNs and nonexistent federations.

For that case, yeah, multiple queries is fine. It's not ideal, but that will make the SQL simpler. I thought you needed the ID. In fact, we already have functions for that, moving one to `dbhelpers/db_helpers.go` wouldn't be a bad idea.

https://github.com/apache/trafficcontrol/blob/master/traffic_ops/traffic_ops_golang/crconfig/servers.go#L366 

Though, you can still get that in a single IMO not-inelegant query

```
select distinct(cdn.name), f.cname, f.description
from cdn
left outer join deliveryservice as d on d.cdn_id = cdn.id
left outer join federation_deliveryservice as fd on fd.deliveryservice = d.id
left outer join federation as f on f.id = fd.federation
where cdn.name = $1;
```

Which will yield 1 row with the CDN and null federation info, if the CDN exists with no federations, and no rows if the CDN doesn't exist.

But, I wouldn't object to either way.

File: traffic_ops/client/v13/cdnfederations.go (hunk 1 above)
Comment (moltzaum) [reply traffic_ops/client/v13/cdnfederations.go]: This is outdated, but I didn't change the direct lines of code. Can you close this manually, or will I need to add a small change to what you selected?

Comment (rawlinp) [reply traffic_ops/client/v13/cdnfederations.go]: Don't worry about getting this thread to collapse into the "Show outdated" thing; you can just reply "done" so everyone knows you addressed the comment.

File: traffic_ops/traffic_ops_golang/routes.go [hunk 4]
```diff
@@ -292,6 +293,13 @@ func Routes(d ServerData) ([]Route, []RawRoute, http.Handler, error) {
 		{1.3, http.MethodPut, `deliveryservices/{xmlID}/urisignkeys$`, saveDeliveryServiceURIKeysHandler, auth.PrivLevelAdmin, Authenticated, nil},
 		{1.3, http.MethodDelete, `deliveryservices/{xmlID}/urisignkeys$`, removeDeliveryServiceURIKeysHandler, auth.PrivLevelAdmin, Authenticated, nil},
 
+		// Federations by CDN (the actual table for `federation')
+		{1.1, http.MethodGet, `cdns/{name}/federations/?(\.json)?$`, api.ReadHandler(cdnfederation.GetTypeSingleton()), auth.PrivLevelReadOnly, Authenticated, nil},
```
Comment (mitchell852) [inline traffic_ops/traffic_ops_golang/routes.go]: 1.1 is correct

**Comment by mitchell852** [comment]
I'm getting this error when trying to hit this in the browser https://localhost:8443/api/1.3/cdns/over-the-top/federations

```
INFO: routing.go:226: 2018-07-19T15:47:02.356745258-06:00: GET /api/1.3/cdns/over-the-top/federations handling (reqid 21)
DEBUG: cdnfederations.go:156: 2018-07-19T15:47:02.359697332-06:00:
--
 Where:
WHERE cdn.name=:name
 Order By:
DEBUG: shared_handlers.go:142: 2018-07-19T15:47:02.359738287-06:00: Query is
	SELECT federation.id as id, cname, ttl, description, ds.id as ds_id, xml_id FROM federation
  JOIN federation_deliveryservice as fd ON federation.id = fd.federation
  JOIN deliveryservice as ds ON ds.id = fd.deliveryservice
	JOIN cdn ON cdn.id = cdn_id
WHERE cdn.name=:name
ERROR: alerts.go:81: 2018-07-19T15:47:02.371098488-06:00: 127.0.0.1:64614 [database access error]
```

**Comment by mitchell852** [comment]
Besides the issue I mentioned above ^^ this seems to work great for all the CRUD functions. I tested these endpoints using TP as the API consumer.

One question. When I create/update a federation, one of the required request attributes is dsId I believe. Does the API check tenancy on this value? If not it should.

File: traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+	if err != nil {
+		log.Errorf("getting tenant list for user: %v\n", err)
+		return nil, []error{tc.DBError}, tc.SystemError
+	}
+
+	var query string
+	_, id := parameters["id"]
+	queryParamsToQueryCols := map[string]dbhelpers.WhereColumnInfo{
+		//db tag                                         symbol from query
+		"id":          dbhelpers.WhereColumnInfo{Column: "federation.id", Checker: api.IsInt},
```
[the remark this answers is not in the document]
Comment (moltzaum) [reply traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: I'm not quite sure what you mean. Yes, it is for parameters, but it is also allows the struct scan  to properly map the rest of the parameters. Looking at origins you include a number of fields that aren't parameters with a comment that says "see the fields mapped in the SQL query".

File: traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+		log.Errorf("unexpected nil id\n")
+	}
+	//Note: the tenantID is not found via a recursive search. The recursive search
+	//is done in IsResourceAuthorizedToUser below.
+	tenantID, err := getTenantIDFromFedID(*fed.ID, fed.ReqInfo.Tx.Tx)
+	if err != nil {
+		//If nobody has claimed a tenant, that federation is publicly visible.
+		//This logically follows /federations/:id/deliveryservices
+		if err == sql.ErrNoRows {
+			log.Errorf("no tenacy")
```
[the remark this answers is not in the document]
Comment (moltzaum) [reply traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: Accidentally left that one in.

File: traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (rawlinp) [reply traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: So the struct scan only needs to "fill in" the federation.id and cdn.name. By adding things like "ds_id", that means you can do `GET /api/1.1/cdns/foo/federations?description=foo` (`descriptions` is the _query parameter_) to get federations where description = "foo". Right now I think we don't need to support anything other than (id -> federation.id, name -> cdn.name).

[the remark this answers is not in the document]
Comment (rawlinp) [reply traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: one of the tricky things here is "path parameters" vs "query parameters". Both `id` and `name` in this API are _path_ parameters (they come from /cdns/{name}/federations/{id} in routes.go), and everything else is a query parameter. The Perl version doesn't support any query parameters, only those two path parameters

File: traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+	}
+	return nil, tc.NoError
+}
+
+//Function not exported because although DELETE and UPDATE have normal tenacy check,
+//CREATE does not. No ds is associated on create. This isn't used for READ because
+//psql doesn't like nested queries.
+func (fed TOCDNFederation) isTenantAuthorized() (bool, error) {
+	if fed.ID == nil {
+		log.Errorf("unexpected nil id\n")
```
[the remark this answers is not in the document]
Comment (moltzaum) [reply traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: It does return an error. The next line is `return nil, []error{tc.DBError}, tc.SystemError`. Maybe clarify?

File: traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+		fed.DeliveryServiceIDs = nil
+	}
+
+	result, err := fed.ReqInfo.Tx.NamedExec(updateQuery(), fed)
+	if err != nil {
+		return parseQueryError(err, "update")
+	}
+
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (moltzaum) [reply traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: done

File: traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (rawlinp) [reply traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: I don't see the line you're talking about. What I see is ```tenantID, err := getTenantIDFromFedID(*fed.ID, fed.ReqInfo.Tx.Tx)``` in which fed.ID could be nil, which would cause a nil pointer dereference panic

[the remark this answers is not in the document]
Comment (moltzaum) [reply traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: I see it now.

File: traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+	}
+	if id {
+		query = selectByID()
+	} else { //searching by name
+		queryParamsToQueryCols["name"] = dbhelpers.WhereColumnInfo{Column: "cdn.name", Checker: nil}
+		query = selectByCDNName()
+	}
+
+	where, orderBy, queryValues, errs := dbhelpers.BuildWhereAndOrderBy(parameters, queryParamsToQueryCols)
+	where, queryValues = dbhelpers.AddTenancyCheck(where, queryValues, "(tenant_id", tenantIDs)
```
[the remark this answers is not in the document]
Comment (moltzaum) [reply traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: ok

File: traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+		}
+
+		federations = append(federations, fed)
+	}
+
+	//if federations yields "response": []
+	if len(federations) == 0 {
+
+		if id {
+			return nil, []error{tc.TenantDSUserNotAuthError}, tc.ForbiddenError
```
[the remark this answers is not in the document]
Comment (moltzaum) [reply traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: Good catch. However, this also covers the case where AddTenancyCheck causes there to be no results and Jeremy told me that tenancy check failures should return a 403. Hmm...

File: traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+	//WHERE cdn.name = :cdn_name (determined by dbhelper)
+}
+
+func updateQuery() string {
+	return `
+	UPDATE federation SET
+	cname=:cname,
+	ttl=:ttl,
+	description=:description
+	WHERE id=:id
```
[the remark this answers is not in the document]
Comment (moltzaum) [reply traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: I was following the [API] (http://traffic-control-cdn.readthedocs.io/en/latest/api/v12/federation.html), where last-updated isn't a part of the response. Since there were other mentions of last-updated from the same resource it seemed like an explicit decision to have it removed from this particular response. Are you saying I should add it now regardless?

[the remark this answers is not in the document]
Comment (rawlinp) [reply traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: So a real problem we have as a project is that devs don't always update the documentation when they update the code. So "Code is King" and should be depended on more than the docs when rewriting existing Perl API endpoints.

Also, something that all our Go APIs are missing from the Perl version is returning all the same stuff on POST/PUT - typically JOINed fields like 'fooName' rather than just 'fooID'. We don't have that exact problem here, but I think we should still be returning `last_updated` here.

File: traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go [hunk 12]
```diff
@@ -0,0 +1,389 @@
...
+		}
+
+		federations = append(federations, fed)
+	}
+
+	// if federations yields "response": []
+	if len(federations) == 0 {
+
+		if id {
+			return nil, []error{tc.TenantDSUserNotAuthError}, tc.ForbiddenError
```
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: So if the request is asking for the federation by ID, and there are no federations with that ID, then the API should return a 404. If it asks by ID and the federation with that ID exists but the tenant doesn't have access to it, the API should return a 401.

If the requests asks for _all_ of the cdn federations (e.g. like the "index" operation), but there are none to return (even when some are filtered out via tenancy), the API should return a 200 with an empty `[]` rather than a 404.

Comment (moltzaum) [reply traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: The second paragraph is currently how the code works. 👍
Now just have to differentiate between the 401 and 404.

File: docs/source/api/v13/federation.rst [hunk 13]
```diff
@@ -0,0 +1,449 @@
+..
```
Comment (rawlinp) [inline docs/source/api/v13/federation.rst]: Since this PR rewrites an existing endpoint with the same functionality, the existing API doc should be updated rather than copying the doc into 1.3 and updating.

I believe 1.3 APIs are for _new_ endpoints that only exist in traffic_ops_golang

**Comment by mitchell852** [comment]
I'm running into problems updating or deleting a federation and it is related to tenancy. I am getting this error:

"user not authorized for requested tenant"

even though i'm logged in as a user with the root tenant...

**Comment by mitchell852** [comment]
I've confirmed all the functionality of these endpoints thru TP.

File: docs/source/api/v12/federation.rst [hunk 14]
```diff
@@ -173,21 +173,24 @@ Federation
...
+            "id": 41
+            "cname": "booya.com.",
+            "ttl": 34,
+            "description": "fooya",
+            "deliveryService": {
+                "id": 61,
+                "xmlId": "the-xml-id"
+            }
+            "lastUpdated": "2018-08-01 14:41:48+00"
+        }
```
Comment (rawlinp) [inline docs/source/api/v12/federation.rst]: this json is missing some commas, but I won't block this for that

File: traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go [hunk 15]
```diff
@@ -0,0 +1,414 @@
...
+	where, orderBy, queryValues, errs := dbhelpers.BuildWhereAndOrderBy(parameters, queryParamsToQueryCols)
+
+	if id {
+		// Can't use `AddTenancyCheck` for id because then we won't know what caused
+		// an empty response, so the tenancy check will be performed below.
+		query = selectByID()
+	} else { // searching by name
+		query = selectByCDNName()
+		where, queryValues = dbhelpers.AddTenancyCheck(where, queryValues, "(tenant_id", tenantIDs)
+		where += " OR tenant_id IS NULL)"
```
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: I think this null tenant_id check can be removed because we now enforce non-null tenancy everywhere, but I won't hold this up because of that.

File: traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go [hunk 16]
```diff
@@ -0,0 +1,414 @@
...
+		return nil, []error{errors.New("resource not found")}, tc.DataMissingError
+	}
+
+	return federations, []error{}, tc.NoError
+}
+
+func (fed *TOCDNFederation) Update() (error, tc.ApiErrorType) {
+
+	if ok, err := fed.isTenantAuthorized(); err != nil {
+		log.Errorf("checking tenacy: %v", err)
```
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: tenacy -> tenancy

File: traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go [hunk 17]
```diff
@@ -0,0 +1,414 @@
...
+
+	return nil, tc.NoError
+}
+
+// In the perl version, :name is ignored. It is not even verified whether or not
+// :name is a real cdn that exists. This mimicks the perl behavior.
+func (fed *TOCDNFederation) Delete() (error, tc.ApiErrorType) {
+
+	if ok, err := fed.isTenantAuthorized(); err != nil {
+		log.Errorf("checking tenacy: %v", err)
```
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: tenacy -> tenancy

File: traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go [hunk 18]
```diff
@@ -0,0 +1,414 @@
...
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func getTenantIDFromFedID(id int, tx *sql.Tx) (int, error) {
+	tenantID := 0
+	query := `
+	SELECT ds.tenant_id from federation as f
```
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: from -> FROM

File: traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go [hunk 19]
```diff
@@ -0,0 +1,414 @@
...
+	JOIN federation_deliveryservice as fd ON f.id = fd.federation
+	JOIN deliveryservice as ds ON ds.id = fd.deliveryservice
+	WHERE f.id = $1`
+	err := tx.QueryRow(query, id).Scan(&tenantID)
+	return tenantID, err
+}
+
+func selectByID() string {
+	return `
+	SELECT tenant_id, federation.id as id, cname, ttl, description, federation.last_updated as last_updated, ds.id as ds_id, xml_id FROM federation
```
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: can you split up each column in the `SELECT` part onto their own lines? And put the `FROM` on its own line as well? Also, please specify which table the column is from (e.g. federation.cname, etc).

File: traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go [hunk 20]
```diff
@@ -0,0 +1,414 @@
...
+	LEFT JOIN federation_deliveryservice as fd ON federation.id = fd.federation
+	LEFT JOIN deliveryservice as ds ON ds.id = fd.deliveryservice`
+	// WHERE federation.id = :id (determined by dbhelper)
+}
+
+func selectByCDNName() string {
+	return `
+	SELECT tenant_id, federation.id as id, cname, ttl, description, federation.last_updated as last_updated, ds.id as ds_id, xml_id FROM federation
```
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go]: above comment applies here too
